docs/source/developer/contributing.rst

```diff
@@ -279,22 +279,71 @@ This translates into the fact that the test suite passes without errors
+continuous integration information for as many combinations as can be supported
+by the service.  Azure CI automatically tests all pull requests on Windows, OS X
+and Linux, as well as a sampling of different Python and NumPy versions. If you
+see problems on platforms you are unfamiliar with, feel free to ask for help in
+your pull request. The Numba core developers can help diagnose cross-platform
+compatibility issues. Also see the :ref:`continuous integration
+<continuous_integration_testing>` section on how public CI is implemented.
+
+.. _continuous_integration_testing:
+
+Continuous integration testing
+''''''''''''''''''''''''''''''
+
+The Numba test suite causes CI systems a lot of grief:
+
+#. It's huge, 9000+ tests.
+#. In part because of 1. and that compilers are pretty involved, the test suite
+   takes a long time to run.
+#. There's sections of the test suite that are deliberately designed to stress
+   systems almost to the point of failure (tests which concurrently compile and
+   execute with threads and fork processes etc).
+#. The combination of things that Numba has to test well exceeds the capacity of
+   any public CI system, (Python versions x NumPy versions x Operating systems
+   x Architectures x feature libraries (e.g. SVML) x threading backends
+   (e.g. OpenMP, TBB)) and then there's CUDA and ROCm too and all their version
+   variants.
+
+As a result of the above, public CI is implemented as follows:
+
+#. The combination of OS x Python x NumPy x Various Features in the testing
+   matrix is designed to give a good indicative result for whether "this pull
+   request is probably ok".
+#. When public CI runs it:
+
+   #. Looks for files that contain tests that have been altered by the proposed
+      change and runs these on the whole testing matrix.
+   #. Runs a subset of the test suite on each part of the testing matrix. i.e.
+      slice the test suite up by the number of combinations in the testing
+      matrix and each combination runs one chunk. This is done for speed,
+      because public CI cannot cope with the load else.
+
+If a pull request is changing CUDA or ROCm code (which cannot be tested on
+Public CI as there's no hardware) or it is making changes to something that the
+core developers consider risky, then it will also be run on the Numba farm just
+to make sure. The Numba project's private build and test farm will actually
+exercise all the applicable tests on all the combinations noted above on real
+hardware!
+
+Helping with Public CI load
+'''''''''''''''''''''''''''
+
+Even with the mitigating design above public CI can get overloaded which causes
+a backlog of builds. It's therefore really helpful when opening pull requests if
+you can limit the frequency of pushing changes. Ideally, please squash commits
+to reduce the number of patches and/or push as infrequently as possible. The
+core developers thank everyone for their cooperation with this!
```

Comment:
Should we remind people not to rebase or force push squashed commits here as it messes up the code review history?

Reply:
Excellent point, yes, will add!

Reply:
Done in 482bdb9